File: backend/app/schemas/creative.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import CreativeAssetStatus, CreativeAssetType, CreativeSetStatus


class CreativeGenerateRequest(BaseModel):
    tone: str | None = Field(default=None, max_length=100)
    objective: str | None = Field(default=None, max_length=500)
    max_headlines: int = Field(default=5, ge=1, le=10)
    max_ad_copy_variants: int = Field(default=3, ge=1, le=5)
    include_targeting: bool = True


class CreativeAssetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    creative_set_id: UUID
    asset_type: CreativeAssetType
    title: str | None
    content: dict[str, Any]
    body_text: str | None
    sort_order: int
    status: CreativeAssetStatus
    reviewed_by_id: UUID | None
    reviewed_at: datetime | None
    rejection_reason: str | None
    created_at: datetime
    updated_at: datetime


class CreativeAssetUpdate(BaseModel):
    status: CreativeAssetStatus | None = None
    rejection_reason: str | None = Field(default=None, max_length=500)


class CreativeSetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    lead_id: UUID
    created_by_id: UUID | None
    signal_id: UUID | None
    scrape_job_id: UUID | None
    scrape_result_id: UUID | None
    status: CreativeSetStatus
    campaign_name: str | None
    campaign_objective: str | None
    generation_params: dict[str, Any]
    extraction_model: str | None
    agent_metadata: dict[str, Any] = Field(default_factory=dict)
    error_message: str | None
    created_at: datetime
    updated_at: datetime
    assets: list[CreativeAssetResponse] = Field(default_factory=list)


class PaginatedCreativeSetsResponse(BaseModel):
    items: list[CreativeSetResponse]
    total: int
    limit: int
    offset: int


class PaginatedCreativeAssetsResponse(BaseModel):
    items: list[CreativeAssetResponse]
    total: int
    limit: int
    offset: int

File: frontend/src/types/creative.ts
export type CreativeSetStatus = 'pending' | 'generating' | 'completed' | 'failed'

export type CreativeAssetType =
  | 'headline'
  | 'ad_copy'
  | 'campaign_idea'
  | 'targeting_suggestion'

export type CreativeAssetStatus = 'draft' | 'approved' | 'rejected'

export interface CreativeAsset {
  id: string
  organization_id: string
  creative_set_id: string
  asset_type: CreativeAssetType
  title: string | null
  content: Record<string, unknown>
  body_text: string | null
  sort_order: number
  status: CreativeAssetStatus
  reviewed_by_id: string | null
  reviewed_at: string | null
  rejection_reason: string | null
  created_at: string
  updated_at: string
}

export interface CreativeSet {
  id: string
  organization_id: string
  lead_id: string
  created_by_id: string | null
  signal_id: string | null
  scrape_job_id: string | null
  scrape_result_id: string | null
  status: CreativeSetStatus
  campaign_name: string | null
  campaign_objective: string | null
  generation_params: Record<string, unknown>
  extraction_model: string | null
  agent_metadata: Record<string, unknown>
  error_message: string | null
  created_at: string
  updated_at: string
  assets: CreativeAsset[]
}

export interface PaginatedCreativeSetsResponse {
  items: CreativeSet[]
  total: number
  limit: number
  offset: number
}

export interface PaginatedCreativeAssetsResponse {
  items: CreativeAsset[]
  total: number
  limit: number
  offset: number
}

export interface CreativeGenerateRequest {
  tone?: string | null
  objective?: string | null
  max_headlines?: number
  max_ad_copy_variants?: number
  include_targeting?: boolean
}

export interface CreativeAssetUpdateRequest {
  status?: CreativeAssetStatus
  rejection_reason?: string | null
}
